ACRA 4.7 - Error 400 while sending report

Hello,

I have setup as described here (https://github.com/ACRA/acralyzer/wiki/setup#configure-your-application-to-send-reports-to-your-acra-storage) acralyzer on cloudant.
I can access the interface and i have my key/pwd pair

Here is the Config
<code>
@ReportsCrashes(
    formUri = "https://xxxxxxx/acra-myapp/_design/acra-storage/_update/report",
    reportType = org.acra.sender.HttpSender.Type.JSON,
    httpMethod = org.acra.sender.HttpSender.Method.PUT,
    formUriBasicAuthLogin="xxxx",
    formUriBasicAuthPassword="xxxx"
)
</code>

I have generated a null pointer to test but every time I get this error saying the response code was 400..

<code>
12-14 01:08:00.107  28212-28212/com..... E/ACRA﹕ ACRA caught a NullPointerException for ,,,,
    java.lang.NullPointerException: Attempt to invoke interface method 'boolean java.util.List.add(java.lang.Object)' on a null object reference
            at com.xxxxx.onCreateView(xxxxx.java:116)
            at android.support.v4.app.Fragment.performCreateView(Fragment.java:1962)
            at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1026)
            at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1207)
            at android.support.v4.app.BackStackRecord.run(BackStackRecord.java:738)
            at android.support.v4.app.FragmentManagerImpl.execPendingActions(FragmentManager.java:1572)
            at android.support.v4.app.FragmentManagerImpl$1.run(FragmentManager.java:493)
            at android.os.Handler.handleCallback(Handler.java:739)
            at android.os.Handler.dispatchMessage(Handler.java:95)
            at android.os.Looper.loop(Looper.java:145)
            at android.app.ActivityThread.main(ActivityThread.java:6837)
            at java.lang.reflect.Method.invoke(Native Method)
            at java.lang.reflect.Method.invoke(Method.java:372)
            at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1404)
            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1199)
12-14 01:08:00.797  28212-29848/? E/ACRA﹕ Failed to send crash report for 1450055211000-approved.stacktrace
    org.acra.sender.ReportSenderException: Error while sending JSON report via Http PUT
            at org.acra.sender.HttpSender.send(HttpSender.java:248)
            at org.acra.SendWorker.sendCrashReport(SendWorker.java:181)
            at org.acra.SendWorker.checkAndSendReports(SendWorker.java:140)
            at org.acra.SendWorker.run(SendWorker.java:76)
     Caused by: java.io.IOException: Host returned error code 400
            at org.acra.util.HttpRequest.send(HttpRequest.java:148)
            at org.acra.sender.HttpSender.send(HttpSender.java:245)
            at org.acra.SendWorker.sendCrashReport(SendWorker.java:181)
            at org.acra.SendWorker.checkAndSendReports(SendWorker.java:140)
            at org.acra.SendWorker.run(SendWorker.java:76)</code>


I have Android 5.1.1
Samsung S6

Same problem on my emulator with version 4.1.2

I too got the same error. Hope my log file can help you in case.

4.7.0 Stacktrace
```
12-18 11:45:20.797 27750-27804/ E/ACRA: Failed to send crash report for 1450419301687-approved.stacktrace
							org.acra.sender.c: Error while sending JSON report via Http POST
								at org.acra.sender.HttpSender.send(SourceFile:248)
								at org.acra.ad.a(SourceFile:181)
								at org.acra.ad.a(SourceFile:140)
								at org.acra.ad.run(SourceFile:76)
							 Caused by: java.io.IOException: Host returned error code 400
								at org.acra.e.b.a(SourceFile:148)
								at org.acra.sender.HttpSender.send(SourceFile:245)
								at org.acra.ad.a(SourceFile:181) 
								at org.acra.ad.a(SourceFile:140) 
								at org.acra.ad.run(SourceFile:76) 

12-18 11:45:01.677 26977-26977/? E/ACRA: Cannot find application log file : ''

```
4.6.2 StackTrace 

```
12-18 12:45:15.797 31737-31737/com.test.myapp D/ACRA: Building report
12-18 12:45:15.797 31737-31737/com.test.myapp D/ACRA: Using custom Report Fields
12-18 12:45:15.837 31737-31737/? I/ACRA: READ_LOGS granted! ACRA can include LogCat and DropBox data.
12-18 12:45:15.837 31737-31737/? E/ACRA: Error while reading application log file 
                                         java.io.FileNotFoundException: /data/data/com.test.myapp/files: open failed: EISDIR (Is a directory)
                                             at libcore.io.IoBridge.open(IoBridge.java:456)
                                             at java.io.FileInputStream.<init>(FileInputStream.java:76)
                                             at android.app.ContextImpl.openFileInput(ContextImpl.java:1061)
                                             at android.content.ContextWrapper.openFileInput(ContextWrapper.java:175)
                                             at org.acra.b.l.a(SourceFile:61)
                                             at org.acra.b.e.a(SourceFile:334)
                                             at org.acra.ErrorReporter.a(SourceFile:5746)
                                             at org.acra.o.a(SourceFile:1142)
                                             at org.acra.ErrorReporter.uncaughtException(SourceFile:428)
                                             at java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:693)
                                             at java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:690)
                                          Caused by: android.system.ErrnoException: open failed: EISDIR (Is a directory)
                                             at libcore.io.IoBridge.open(IoBridge.java:446)
                                             at java.io.FileInputStream.<init>(FileInputStream.java:76) 
                                             at android.app.ContextImpl.openFileInput(ContextImpl.java:1061) 
                                             at android.content.ContextWrapper.openFileInput(ContextWrapper.java:175) 
                                             at org.acra.b.l.a(SourceFile:61) 
                                             at org.acra.b.e.a(SourceFile:334) 
                                             at org.acra.ErrorReporter.a(SourceFile:5746) 
                                             at org.acra.o.a(SourceFile:1142) 
                                             at org.acra.ErrorReporter.uncaughtException(SourceFile:428) 
                                             at java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:693) 
                                             at java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:690) 
12-18 12:45:15.837 31737-31737/? D/ACRA: Writing crash report file 1450422915000.stacktrace.
12-18 12:45:15.837 31737-31737/? D/ACRA: About to start ReportSenderWorker from #handleException
12-18 12:45:15.847 31737-663/? D/ACRA: Mark all pending reports as approved.
12-18 12:45:15.847 31737-663/? D/ACRA: Looking for error files in /data/data/com.test.myapp/files
12-18 12:45:15.847 31737-664/? D/ACRA: Waiting for 2000 millis from null currentMillis=1450422915852
12-18 12:45:15.847 31737-663/? D/ACRA: #checkAndSendReports - start
12-18 12:45:15.847 31737-663/? D/ACRA: Looking for error files in /data/data/com.test.myapp/files
12-18 12:45:15.847 31737-663/? I/ACRA: Sending file 1450422915000-approved.stacktrace
12-18 12:45:15.847 31737-665/? D/ACRA: Waiting for Toast
12-18 12:45:15.847 31737-31747/? I/art: Debugger is no longer active
12-18 12:45:15.847 31737-663/? D/ACRA: Connect to https://myapp.cloudant.com/acra-myapp/_design/acra-storage/_update/report
12-18 12:45:15.867 31737-656/? E/HardwareRenderer: setHWUICacheConfig must be called in UI thread. Skip!
12-18 12:45:15.877 31737-663/? W/ACRA: Using default httpsSocketFactoryFactory - Could not construct : org.acra.util.DefaultHttpsSocketFactoryFactory
12-18 12:45:15.887 31737-663/? D/ACRA: Sending request to https://myapp.cloudant.com/acra-myapp/_design/acra-storage/_update/report
12-18 12:45:15.887 31737-663/? D/libc: [NET] android_getaddrinfofornet+,hn 17(0x726163652e636c),sn(),hints(known),family 0,flags 4
12-18 12:45:15.887 31737-663/? D/libc: [NET] android_getaddrinfofornet-, err=8
12-18 12:45:15.887 31737-663/? D/libc: [NET] android_getaddrinfofornet+,hn 17(0x726163652e636c),sn(),hints(known),family 0,flags 1024
12-18 12:45:15.887 31737-663/? D/libc: [NET] android_getaddrinfofornet-, pass to proxy
12-18 12:45:15.887 31737-663/? D/libc: [NET] android_getaddrinfo_proxy+
12-18 12:45:15.887 31737-663/? D/libc: [NET] android_getaddrinfo_proxy get netid:0
12-18 12:45:17.327 31737-663/? D/libc: [NET] android_getaddrinfo_proxy-, success
12-18 12:45:17.327 31737-663/? D/libc: [NET] android_getaddrinfofornet+,hn 14(0x3135392e323533),sn(),hints(known),family 0,flags 4
12-18 12:45:17.327 31737-663/? D/libc: [NET] android_getaddrinfofornet-, SUCCESS
12-18 12:45:17.947 31737-665/? D/ACRA: Finished waiting for Toast
12-18 12:45:17.947 31737-665/? D/ACRA: Waiting for Worker
12-18 12:45:20.267 31737-663/? D/libc: [NET] android_getaddrinfofornet+,hn 17(0x726163652e636c),sn(),hints(known),family 0,flags 4
12-18 12:45:20.267 31737-663/? D/libc: [NET] android_getaddrinfofornet-, err=8
12-18 12:45:20.267 31737-663/? D/libc: [NET] android_getaddrinfofornet+,hn 17(0x726163652e636c),sn(),hints(known),family 0,flags 1024
12-18 12:45:20.267 31737-663/? D/libc: [NET] android_getaddrinfofornet-, pass to proxy
12-18 12:45:20.267 31737-663/? D/libc: [NET] android_getaddrinfo_proxy+
12-18 12:45:20.267 31737-663/? D/libc: [NET] android_getaddrinfo_proxy get netid:0
12-18 12:45:20.267 31737-663/? D/libc: [NET] android_getaddrinfo_proxy-, success
12-18 12:45:23.017 31737-663/? V/p: Setting allowed TLS protocols: TLSv1, TLSv1.1, TLSv1.2
12-18 12:45:23.017 31737-663/? D/p: Using documented SNI with host name myapp.cloudant.com
12-18 12:45:25.967 31737-663/? I/p: Established TLSv1.2 connection with myapp.cloudant.com using TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
12-18 12:45:27.927 31737-663/? D/ACRA: #checkAndSendReports - finish
12-18 12:45:28.017 31737-665/? D/ACRA: Finished waiting for Worker
12-18 12:45:28.017 31737-665/? D/ACRA: Wait for Toast + worker ended. Kill Application ? true
```

Same problem code 400 on 4.7, works on 4.6
Device: Nexus 4
Android: 5.1.1

From differences I noticed is additional "\n" near the end of String named encoded when using 4.7. It is not something added during copying and pasting from debugger, it is actual line feed character in char array.

4.6 headers:
```

"Authorization: Basic ***************************************************************************************="
"User-Agent: Android"
"Accept: text/html,application/xml,application/json,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5"
"Content-Type: application/json"
```

4.7 headers:
```
"Accept" "text/html,application/xml,application/json,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5"
"Authorization" "Basic ****************************************************************************\n***********="
"Content-Type" "application/json"
"User-Agent" "Android"
```

4.6 logcat:
```
E/ACRA: ACRA caught a NullPointerException for **************************.android
        java.lang.NullPointerException: Attempt to invoke virtual method 'int java.lang.String.length()' on a null object reference
            at ********************************.create(MainGame.java:50)
            at com.badlogic.gdx.backends.android.AndroidGraphics.onSurfaceChanged(AndroidGraphics.java:243)
            at android.opengl.GLSurfaceView$GLThread.guardedRun(GLSurfaceView.java:1511)
            at android.opengl.GLSurfaceView$GLThread.run(GLSurfaceView.java:1239)
D/ACRA: Building report
D/ACRA: Using custom Report Fields
I/ACRA: READ_LOGS granted! ACRA can include LogCat and DropBox data.
D/ACRA: Writing crash report file 1450813947000.stacktrace.
D/ACRA: About to start ReportSenderWorker from #handleException
D/ACRA: Mark all pending reports as approved.
D/ACRA: Looking for error files in /data/data/**************************.android/files
D/ACRA: #checkAndSendReports - start
D/ACRA: Looking for error files in /data/data/**************************.android/files
I/ACRA: Sending file 1450813947000-approved.stacktrace
D/ACRA: Waiting for Toast
D/ACRA: Waiting for 2000 millis from 1450813947836 currentMillis=1450813947935
D/ACRA: Connect to https://************.cloudant.com/acra-*********/_design/acra-storage/_update/report
D/ACRA: Sending request to https://************.cloudant.com/acra-*********/_design/acra-storage/_update/report/4ff90d01-479b-48fd-a417-2334fdef600b
D/ACRA: #checkAndSendReports - finish
D/ACRA: Finished waiting for Toast
D/ACRA: Waiting for Worker
D/ACRA: Finished waiting for Worker
D/ACRA: Wait for Toast + worker ended. Kill Application ? true
E/ACRA: **************************.android fatal error : Attempt to invoke virtual method 'int java.lang.String.length()' on a null object reference
        java.lang.NullPointerException: Attempt to invoke virtual method 'int java.lang.String.length()' on a null object reference
            at ********************************.create(MainGame.java:50)
            at com.badlogic.gdx.backends.android.AndroidGraphics.onSurfaceChanged(AndroidGraphics.java:243)
            at android.opengl.GLSurfaceView$GLThread.guardedRun(GLSurfaceView.java:1511)
            at android.opengl.GLSurfaceView$GLThread.run(GLSurfaceView.java:1239)
I/ACRA: Finishing the last Activity prior to killing the Process
I/ACRA: Finished class **************************.android.AndroidLauncher
```

4.7 logcat:
```
E/ACRA: ACRA caught a NullPointerException for **************************.android
        java.lang.NullPointerException: Attempt to invoke virtual method 'int java.lang.String.length()' on a null object reference
            at ********************************.create(MainGame.java:50)
            at com.badlogic.gdx.backends.android.AndroidGraphics.onSurfaceChanged(AndroidGraphics.java:243)
            at android.opengl.GLSurfaceView$GLThread.guardedRun(GLSurfaceView.java:1511)
            at android.opengl.GLSurfaceView$GLThread.run(GLSurfaceView.java:1239)
D/ACRA: Building report
D/ACRA: Using custom Report Fields
I/ACRA: READ_LOGS granted! ACRA can include LogCat and DropBox data.
D/ACRA: Writing crash report file 1450809859203.stacktrace.
D/ACRA: About to start ReportSenderWorker from #handleException
D/ACRA: Mark all pending reports as approved.
D/ACRA: Looking for error files in /data/data/**************************.android/files
D/ACRA: Waiting for Toast
D/ACRA: Waiting for 2000 millis from 1450809859136 currentMillis=1450809859217
D/ACRA: #checkAndSendReports - start
D/ACRA: Looking for error files in /data/data/**************************.android/files
I/ACRA: Sending file 1450809809030-approved.stacktrace
D/ACRA: Sending report using org.acra.sender.HttpSender
D/ACRA: Connect to https://************.cloudant.com/acra-*********/_design/acra-storage/_update/report
D/ACRA: Sending request to https://************.cloudant.com/acra-*********/_design/acra-storage/_update/report/af4f0e92-0109-45ee-a833-f9d87fcd9bae
D/ACRA: Request response : 400 : Bad request
E/ACRA: Failed to send crash report for 1450809809030-approved.stacktrace
        org.acra.sender.ReportSenderException: Error while sending JSON report via Http PUT
            at org.acra.sender.HttpSender.send(HttpSender.java:248)
            at org.acra.SendWorker.sendCrashReport(SendWorker.java:181)
            at org.acra.SendWorker.checkAndSendReports(SendWorker.java:140)
            at org.acra.SendWorker.run(SendWorker.java:76)
         Caused by: java.io.IOException: Host returned error code 400
            at org.acra.util.HttpRequest.send(HttpRequest.java:148)
            at org.acra.sender.HttpSender.send(HttpSender.java:245)
            at org.acra.SendWorker.sendCrashReport(SendWorker.java:181) 
            at org.acra.SendWorker.checkAndSendReports(SendWorker.java:140) 
            at org.acra.SendWorker.run(SendWorker.java:76) 
D/ACRA: #checkAndSendReports - finish
D/ACRA: Sending request to https://************.cloudant.com/acra-*********/_design/acra-storage/_update/report/af4f0e92-0109-45ee-a833-f9d87fcd9bae
D/ACRA: Request response : 400 : Bad request
E/ACRA: Failed to send crash report for 1450809809030-approved.stacktrace
        org.acra.sender.ReportSenderException: Error while sending JSON report via Http PUT
            at org.acra.sender.HttpSender.send(HttpSender.java:248)
            at org.acra.SendWorker.sendCrashReport(SendWorker.java:181)
            at org.acra.SendWorker.checkAndSendReports(SendWorker.java:140)
            at org.acra.SendWorker.run(SendWorker.java:76)
         Caused by: java.io.IOException: Host returned error code 400
            at org.acra.util.HttpRequest.send(HttpRequest.java:148)
            at org.acra.sender.HttpSender.send(HttpSender.java:245)
            at org.acra.SendWorker.sendCrashReport(SendWorker.java:181) 
            at org.acra.SendWorker.checkAndSendReports(SendWorker.java:140) 
            at org.acra.SendWorker.run(SendWorker.java:76) 
I/ACRA: Sending file 1450809859203-approved.stacktrace
D/ACRA: Sending report using org.acra.sender.HttpSender
D/ACRA: Connect to https://************.cloudant.com/acra-*********/_design/acra-storage/_update/report
D/ACRA: Sending request to https://************.cloudant.com/acra-*********/_design/acra-storage/_update/report/9ba7994c-fc1d-4f87-86e9-d6c7d3f0ff82
D/ACRA: Request response : 400 : Bad request
E/ACRA: Failed to send crash report for 1450809859203-approved.stacktrace
        org.acra.sender.ReportSenderException: Error while sending JSON report via Http PUT
            at org.acra.sender.HttpSender.send(HttpSender.java:248)
            at org.acra.SendWorker.sendCrashReport(SendWorker.java:181)
            at org.acra.SendWorker.checkAndSendReports(SendWorker.java:140)
            at org.acra.SendWorker.run(SendWorker.java:76)
         Caused by: java.io.IOException: Host returned error code 400
            at org.acra.util.HttpRequest.send(HttpRequest.java:148)
            at org.acra.sender.HttpSender.send(HttpSender.java:245)
            at org.acra.SendWorker.sendCrashReport(SendWorker.java:181) 
            at org.acra.SendWorker.checkAndSendReports(SendWorker.java:140) 
            at org.acra.SendWorker.run(SendWorker.java:76) 
D/ACRA: #checkAndSendReports - finish
D/ACRA: Finished waiting for Toast
D/ACRA: Waiting for Worker
D/ACRA: Finished waiting for Worker
D/ACRA: Wait for Toast + worker ended. Kill Application ? true
E/ACRA: **************************.android fatal error : Attempt to invoke virtual method 'int java.lang.String.length()' on a null object reference
        java.lang.NullPointerException: Attempt to invoke virtual method 'int java.lang.String.length()' on a null object reference
            at ********************************.create(MainGame.java:50)
            at com.badlogic.gdx.backends.android.AndroidGraphics.onSurfaceChanged(AndroidGraphics.java:243)
            at android.opengl.GLSurfaceView$GLThread.guardedRun(GLSurfaceView.java:1511)
            at android.opengl.GLSurfaceView$GLThread.run(GLSurfaceView.java:1239)
I/ACRA: Finishing the last Activity prior to killing the Process
I/ACRA: Finished class **************************.android.AndroidLauncher
```

--- a/src/main/java/org/acra/util/HttpRequest.java
+++ b/src/main/java/org/acra/util/HttpRequest.java
@@ -89,7 +89,7 @@ public final class HttpRequest {
         // Set Credentials
         if ((login != null) && (password != null)) {
             final String credentials = login + ":" + password;
-            final String encoded = new String(Base64.encode(credentials.getBytes("UTF-8"), Base64.DEFAULT), "UTF-8");
+            final String encoded = new String(Base64.encode(credentials.getBytes("UTF-8"), Base64.NO_WRAP), "UTF-8");
             urlConnection.setRequestProperty("Authorization", "Basic " + encoded);
         }
 

